viamonvif/discovery.go

```diff
@@ -0,0 +1,308 @@
+	}
+	logger.Debugf("GetDeviceInformation response body: %s", deviceInfoBody)
+	deviceInfoResponse.Body = io.NopCloser(bytes.NewReader(deviceInfoBody))
+
+	var deviceInfoEnvelope getDeviceInformationResponse
+	err = xml.NewDecoder(deviceInfoResponse.Body).Decode(&deviceInfoEnvelope)
+	if err != nil {
+		return CameraInfo{}, fmt.Errorf("failed to decode device information response: %w", err)
+	}
+
+	// Call the ONVIF Media service to get the available media profiles using the same device instance
+	rtspURLs, err := getRTSPStreamURLs(deviceInstance, username, password, logger)
+	if err != nil {
+		return CameraInfo{}, fmt.Errorf("failed to get RTSP URLs: %w", err)
+	}
+
+	cameraInfo := CameraInfo{
+		RTSPURLs:     rtspURLs,
+		Manufacturer: deviceInfoEnvelope.Body.GetDeviceInformationResponse.Manufacturer,
+		Model:        deviceInfoEnvelope.Body.GetDeviceInformationResponse.Model,
+		SerialNumber: deviceInfoEnvelope.Body.GetDeviceInformationResponse.SerialNumber,
+	}
+
+	return cameraInfo, nil
+}
+
+// getRTSPStreamURLs uses the ONVIF Media service to get the RTSP stream URLs for all available profiles.
+func getRTSPStreamURLs(deviceInstance OnvifDevice, username, password string, logger logging.Logger) ([]string, error) {
+	getProfiles := media.GetProfiles{}
+	profilesResponse, err := deviceInstance.CallMethod(getProfiles)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get media profiles: %w", err)
+	}
+	defer profilesResponse.Body.Close()
+
+	profilesBody, err := io.ReadAll(profilesResponse.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read profiles response body: %w", err)
+	}
+	logger.Debugf("GetProfiles response body: %s", profilesBody)
+	// Reset the response body reader after logging
+	profilesResponse.Body = io.NopCloser(bytes.NewReader(profilesBody))
+
+	var envelope getProfilesResponse
+	err = xml.NewDecoder(profilesResponse.Body).Decode(&envelope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode media profiles response: %w", err)
+	}
+
+	if len(envelope.Body.GetProfilesResponse.Profiles) == 0 {
+		logger.Warn("No media profiles found in the response")
+		return nil, errors.New("no media profiles found")
+	}
+
+	logger.Debugf("Found %d media profiles", len(envelope.Body.GetProfilesResponse.Profiles))
+	for i, profile := range envelope.Body.GetProfilesResponse.Profiles {
+		logger.Debugf("Profile %d: Token=%s, Name=%s", i, profile.Token, profile.Name)
+	}
+
+	// Resultant slice of RTSP URIs
+	var rtspUris []string
+	// Iterate over all profiles and get the RTSP stream URI for each one
+	for _, profile := range envelope.Body.GetProfilesResponse.Profiles {
+		logger.Debugf("Using profile token: %s", profile.Token)
+
+		getStreamURI := media.GetStreamUri{
+			StreamSetup: onvifxsd.StreamSetup{
+				Stream: onvifxsd.StreamType(streamTypeRTPUnicast),
+				Transport: onvifxsd.Transport{
+					Protocol: onvifxsd.TransportProtocol(transportProtocolRTSP),
```

Comment:
Just to clarify here:
https://github.com/use-go/onvif/blob/master/docs/wsdl/media2.wsdl#L679-L686

My understanding here is that we are checking for URIs that have `RTSP (RTP over TCP) support`.

Should we also check for `RtspUnicast (RTP over UDP) support`?

Reply:
Talked in IRL - does not look like we even need to add `Protocol` to query here.

RTSP over UDP vs TCP is another discussion all together. Does the viamrtsp `gortsplib` dep even support udp streaming? Not sure yet.